dmutils/email/dm_notify.py

```diff
@@ -0,0 +1,123 @@
+# -*- coding: utf-8 -*-
+"""Digital Marketplace Notify integration."""
+from collections import OrderedDict
+
+from flask import current_app
+from notifications_python_client import NotificationsAPIClient
+from notifications_python_client.errors import HTTPError
+
+from dmutils.email.exceptions import DMNotifyError
+from dmutils.email.helpers import hash_string
+from dmutils.formats import stringify_dict
+
+
+class DMNotifyClient(object):
+    """Digital Marketplace wrapper around the Notify python client."""
+
+    _client_class = NotificationsAPIClient
+    _sent_references_cache = None
+
+    def __init__(
+            self,
+            govuk_notify_api_key,
+            govuk_notify_base_url='https://api.notifications.service.gov.uk',
+            logger=None
+    ):
+        """Set up logging and mail client."""
+        self.logger = logger or current_app.logger
+        self.client = self._client_class(govuk_notify_api_key, govuk_notify_base_url)
+
+    def get_all_notifications(self, **kwargs):
+        """Wrapper for notifications_python_client.notifications.NotificationsAPIClient::get_all_notifications"""
+        return self.client.get_all_notifications(**kwargs)['notifications']
+
+    def get_delivered_notifications(self, **kwargs):
+        """Wrapper for notifications_python_client.notifications.NotificationsAPIClient::get_all_notifications"""
+        return self.get_all_notifications(status='delivered', **kwargs)
+
+    def get_delivered_references(self, invalidate_cache=False):
+        """Get the references of all notifications that have already been delivered."""
+        if invalidate_cache:
+            self._sent_references_cache = None
+        if self._sent_references_cache is None:
```

Comment:
We need the second line because `_sent_references_cache` can be set to `None` if the `invalidate_cache` flag is set and if the cache has never been instantiated.

It would be possible to use an `or` statement like so:
```
if self._sent_references_cache is None or invalidate_cache:
    self._sent_references_cache  = ...
```

I chose to write it like the above because I think it's more readable. 'We've told it to invalidate the cache, so we do that, then we reset the cache.' Rather than implicitly invalidating the cache by overwriting it.